lib/ansible/modules/cloud/amazon/aws_s3.py

```diff
@@ -501,14 +500,22 @@ def upload_s3file(module, s3, bucket, obj, src, expiry, metadata, encrypt, heade
             extra['ContentType'] = content_type
 
         s3.upload_file(Filename=src, Bucket=bucket, Key=obj, ExtraArgs=extra)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg="Unable to complete PUT operation.")
+    try:
         for acl in module.params.get('permission'):
             s3.put_object_acl(ACL=acl, Bucket=bucket, Key=obj)
+    except s3.exceptions.from_code('XNotImplemented'):
```

Comment:
I assume this should also be added to put_bucket_acl. The only thing I'm wondering about is if we should handle this for other S3 drop-ins as well. I assume most of them will fail for these two calls and all have their own special exception that they raise. Minio, for instance, raising a `NotImplemented` error.

Reply:
Actually, since XNotImplemented is not an exception botocore will recognize mode of the time so looking at the botocore code, it will catch all ClientErrors https://github.com/boto/botocore/blob/cac78632cabddbc7b64f63d99d419fe16917e09b/botocore/errorfactory.py#L40-L42 and will end up mask exceptions from lacking put ACL permissions. I'm going to revert my commit to handle these the old way with ['Error']['Code'].